### PR TITLE
Completed Assessment

### PR DESCRIPTION
**Pull request by mikerogers04**
# Submitting Week 5 Assessment

### ✅ Completed Files

Mark the following as confirmation that the files in this repository are complete.

- [✅] JavaScript Coding Challenges
- [✅] Ruby Coding Challenges
- [✅] Interview Questions

### 📝 Feedback

If you would like specific feedback on a particular problem please leave details here:
Please provide feedback on interview answers. Also, if possible could provide additional context on Coding Challenges 3b. Particularly my original function that is commented out. Is there a way to make that function dynamic instead of hard coding in static numbers(5 and 3) inside the reduce methods on lines 187 and 188.

### CODE REVIEW

File: code-challenges.test.js [hunk 1]
```diff
@@ -21,8 +28,67 @@ const secretCodeWord2 = "Gobbledygook"
...
+        i: '1',
+        o: '0',
+        A: '4',
+        E: '3',
+        I: '1',
+        O: '0'
+    }
+    //Setup regex for lower/upper case letter with global search flag and i flag to ignore casing upon search
+    let regex = /[aeioAEIO]/gi
+    //.replace method to execute on the input string, comparing the regex pattern to the key-value pairs in the object
```
Comment (Genemichael25) [inline code-challenges.test.js:77]: Love this approach! Nicely done 💯

File: interview-questions.md [hunk 2]
```diff
@@ -8,44 +8,83 @@ Then, research the question to expand on your answer. Even if you feel you have
...
 
 2. What is a gem?
 
-Your answer:
+Your answer: A gem is a package manager designed for Ruby. RSpec is an example of a Ruby gem. When working with RSpec, you have to install that gem in the place you're working with Ruby code you want to test. Without installation, you cannot use the testing environment of RSpec on Ruby files. 
 
-Researched answer:
+Researched answer: Further research made it more clear that a gem is not a package manager. While it can be considered a package, the specific package manager for Ruby is called RubyGems. A gem is a library of code functionality that can be imported to be used for developers. To dive a little bit deeper, comparitively in JS, Jest is a testing framework that when installed brings a large amount of dependencies into a single place. That folder being called node_nodules. In Ruby, when using RSpec, you have to import those dependencies onto the test file itself.
```
Comment (Genemichael25) [inline interview-questions.md:37]: Great follow-up! You will definitely work with a few gems here, as you pointed out, RSpec being one of them. They are great in that they provide functionality depending on what they are for.

File: interview-questions.md [hunk 3]
```diff
@@ -8,44 +8,83 @@ Then, research the question to expand on your answer. Even if you feel you have
...
+
+Researched answer: Upon further research, the keyword table was what I was missing in my explanation. A database can have many tables that all have relationships with one another. The collection of all entries in a table that contain the same attributes is a relation. Other types of databases include hierarchical, non-relational, object-oriented, and many more. 
 
 4. What are primary keys? Why are they important?
 
-Your answer:
+Your answer: Primary keys are unique identifiers to individual databases when created. Primary keys are created automatically after database creation. Primary keys go in sequential order (1,2..etc) for every new piece of information added to the database. Primary keys are incredibly important because they help identify every entry into a database. No two entries can have the same primary key. 
 
-Researched answer:
+Researched answer: Primary keys are denoted by id. Primary keys provide a unique identifier to each entry even if those entries have the same information.
```
Comment (Genemichael25) [inline interview-questions.md:58]: Exactly, and by having that primary key we can reference the entry we need even if its data is similar to another one. For example, can you imagine if we had a table with students and we happened to have 3 students named John Smith. Of course there might be other information in that table that would differentiate, but the primary key is a great tool especially if we need to associate them with another table.

File: interview-questions.md [hunk 4]
```diff
@@ -8,44 +8,83 @@ Then, research the question to expand on your answer. Even if you feel you have
...
+Update: Put, Patch
+Delete: Delete
+
+Researched answer: Post is used to create an item. Get is used to retrieve a single item or list of items. Put/Patch is used to update an item. Delete is simply to delete an item. 
+
+HTTP Verb with corresponding status codes:
+Post: 201 sucess, 404/409 fail
+Get: 200 sucess, 404 fail
+Put/Patch: 200, 204 success, 404 fail
+Delete: 200 success, 404 fail
```
Comment (Genemichael25) [inline interview-questions.md:76]: I love the addition of the status codes here! They are absolutely important to keep in mind.

File: ruby-challenges.rb [hunk 5]
```diff
@@ -4,24 +4,87 @@
...
+
+# Pseudo:
+# Create a method called contains_one_letter
+#Input: An array of words and a single letter
+#Use .select to iterate over the array
+#Inside select method, use .include method to check each element in the array if it contains the input letter
+#Output: An array containing the words based on the input letter
+
+def contains_one_letter(array, letter)
+    #Use .select to filter the input array based on the .include condition
```
Comment (Genemichael25) [inline ruby-challenges.rb:24]: Awesome, clean and concise! 👏🏽

File: ruby-challenges.rb [hunk 6]
```diff
@@ -4,24 +4,87 @@
...
+#Create a method called one_ordered_array
+#Input: A hash containing three arrays
+#Use .flatten method on the hash to return a single array with all key-pair values at each index
+#Destructure array then push the array elements only 
+#Refactored portion: Instead of destructuring array, remembered you could call upon .values to grab only values inside a hash. Then used .flatten, then .sort for final result. 
+#Output: A single, alphabetical ordered array
+
+def one_ordered_array(hash)
+    #Refactored to single line to do all data manipulation
+    new_array = hash.values.flatten.sort()
```
Comment (Genemichael25) [inline ruby-challenges.rb:51]: Do we need () after sort?

File: ruby-challenges.rb [hunk 7]
```diff
@@ -4,24 +4,87 @@
...
+        "The #{@model} bike has #{@wheels} wheels and is going #{@current_speed} mph."
+    end
+
+    def pedal_faster(value)
+        @current_speed += value
+    end
+
+    def brake(value)
+        #.max is needed in order to prevent negative numbers upon calling the brake method
+        @current_speed = [@current_speed - value, 0].max
```
Comment (Genemichael25) [inline ruby-challenges.rb:81]: Great! I like the use of max here. Crushed this section!

File: code-challenges.test.js [hunk 8]
```diff
@@ -48,5 +149,98 @@ const hand3 = [5, 5, 5, 5, 4]
...
+    //Count will be incremented based on how many times five and three are in the array
+    //Store those count values into variables
+    //Use if/else conditional to verify if the count values strictly equal three or two(3 for three of a kind and 2 for one pair)
+        //If true, return true. Else, return false
+//Output: A true or false value if the array is a "full house"
+
+// const fullHouse = (array) => {
+
+//     let three = array.reduce((count, threeOfAKind) => threeOfAKind === 5 ? count + 1 : count, 0)
+//     let two = array.reduce((count, twoPair) => twoPair === 3 ? count + 1 : count, 0)
```
Comment (Genemichael25) [inline code-challenges.test.js:188]: I will get back to you on this, let me experiment with this one a bit more.

**Review by Genemichael25 — APPROVED** [review]
Outstanding work Mike! This assessment displays a Competent level of comprehension without a doubt. The interview questions are answered extremely well and I love the examples you are using. Your pseudocode and functions are clean and I appreciate the different approaches you attempt. Overall, you are doing fantastic on these assessments and keep it up, just one more to go; finish off strong! 👏🏽

File: code-challenges.test.js (hunk 8 above)
Comment (mikerogers04) [reply code-challenges.test.js:188]: Looking forward to your follow-up! Feel free to send me a message on Slack about it whenever you have free time!

**Pull request merged**